st2actions/st2actions/workflows/workflows.py

```diff
@@ -85,6 +93,53 @@ def process(self, message):
             # the database and fail the workflow execution gracefully. In this case,
             # the garbage collector will find and cancel these workflow executions.
             self.fail_workflow_execution(message, e)
+        finally:
+            if type(message) == wf_db_models.WorkflowExecutionDB:
+                self._running_workflows.remove(str(message["action_execution"]))
+            else:
+                try:
+                    self._running_workflows.remove(str(message["parent"]))
+                except ValueError:
+                    pass
+
+    def shutdown(self):
+        super(WorkflowExecutionHandler, self).shutdown()
+
+        if cfg.CONF.workflow_engine.graceful_shutdown:
+
+            coordinator = coordination.get_coordinator()
+            member_ids = []
+            service = "workflow_engine"
+            exit_timeout = cfg.CONF.workflow_engine.exit_timeout
+            sleep_delay = cfg.CONF.workflow_engine.sleep_delay
+            timeout = 0
+
+            while timeout < exit_timeout and self._running_workflows:
```

Comment:
Can you put some comments what this while loop achieves?  Some work is still needed so the mechanism is more robust. I also think support of some edge cases is missing (i.e. what happens if all engines shutdown at the same time. will some workflows in running state while other in pause state).